### DISCUSSION TITLE
JWT maximum size?

### DISCUSSION THREAD

**Discussion by tvogel** [supabase/supabase#12057] [Questions] [answered]
Hi!

I am using `app_metadata` to manage user authorization for specific database objects. Now, depending on the granularity of control and trying to limit necessary indirections/joins, the number of "claims" in the `app_metadata` might get large quickly, so I am wondering up to which number of claims, this will be a feasible solution. My `app_metadata` has this structure:

```JSON
{
  "provider": "email"
, "providers": ["email"]
, "authorization": {
    "read": [
      "1717ea5a-8fee-4b2e-8e91-92c27d71e9c2"
    , "e0dc6621-200e-4dff-9d16-8f4297d22b20"
    ]
  , "write": [
      "1717ea5a-8fee-4b2e-8e91-92c27d71e9c2"
    ]
  }
}
```
where each UUID identifies some database object.

So, is there any hard or practical limit?

**Comment by GaryAustin1** [comment, marked as answer]
I don't know if there is a hard limit on a jwt, but that seems like a bad idea.

There are, I am pretty sure limits, on http header sizes and the jwt is passed in headers. https://community.cloudflare.com/t/maximum-on-http-header-values/424067

Every request has to carry around the extra data and all your token refreshes will need to return that extra data.  

The tradeoff as you say is using joins in your policies versus processing a bunch of json in a function (although I guess one could do a string search also).

**Comment by MentalGear** [comment]
What route did you end up going? @tvogel 

I was also thinking of taking the "data in Supabase JWT" route, but having to manually refresh the JWT each time for frequent new data is not a good idea, as @GaryAustin1 was saying. (e.g. other tabs with your website will still have the old token, leading to potential problems)

It seems like a ["custom header"](https://developers.cloudflare.com/workers/examples/auth-with-headers/) approach, whereas a supabase edge function returns the query along with the encrypted header that has the object names which the client request can access. This header is then passed again for every object request (from the client ) to the edge function (which has the signing key to decrypt/validate the header) that controls permissions/access. In this way, I only need to do one table lookup (the initial query for the objects), and no more lengthy database-lookups for each object request.

Example: Your SQL query returns x elements along with an encrypted header with the UUID of each of those objects. The encrypted header is passed from the client to the object request function, which can verify it, and then send back the request object if it is contained in the 'encrypted header '.

I might even use a serial id for my `files` table and storage name, so I can save on header space (e.g. `1` vs `1e1d0985-2269-4bb3-a7b9-c632bf19e673` which can potentially add up).

**Reply by GaryAustin1** [reply]
There is a new feature for hooking the jwt generation which might be useful for getting faster updates of the jwt.
https://supabase.com/docs/guides/auth/custom-claims-and-role-based-access-control-rbac
Does not address the concern though of jwt size for this particular approach of lots of permissions.

**Reply by tvogel** [reply]
I went the route to not define authorization on a per-item level (stored in the JWT) but on a per-organization level (stored in the JWT), so there are more JOINs necessary for calculating visibility but the JWT is just not the right place to represent this level of granularity.
